src: added big-endian check and support to UTF-16 encode

exinfinitum opened the pull request:
Added for loop to Encode function to support big-endian machines.
Used stack allocation instead of heap allocation to improve speed of endianness correction.
Resolves https://github.com/nodejs/node/issues/3409

jasnell commented:
@srl295

bnoordhuis commented on src/string_bytes.cc:
```diff
@@ -857,7 +857,17 @@ Local<Value> StringBytes::Encode(Isolate* isolate,
                                  size_t buflen) {
   Local<String> val;
-
+  uint16_t dst[buflen];
```
Don't do this.  `buflen` may be on the order of 100s of megabytes.

bnoordhuis commented:
Instead of copy/pasting the byte-swapping code, it would be much better to move it into a small utility class.

kkoopa commented on src/string_bytes.cc:
```diff
@@ -857,7 +857,17 @@ Local<Value> StringBytes::Encode(Isolate* isolate,
+    // http://nodejs.org/api/buffer.html regarding Node's "ucs2"
+    // encoding specification
+    for (size_t i = 0; i < buflen; i++) {
+      dst[i] = (buf[i] << 8) | (buf[i] >> 8);
```
Would it be worth using intrinsics to ensure efficient byteswap instructions?

bnoordhuis replied:
There's no need.  gcc and clang recognize the byte swap just fine.

jasnell commented:
agree with @bnoordhuis ... there are a few places in the code now where we do this kind of reordering. We should be refactoring it down rather than cutting and pasting it elsewhere.

jasnell commented:
@srl295 @bnoordhuis ...

@exinfinitum ... just an fyi... we don't get notified automatically when a PR has been updated with a new push. It's helpful if you comment and mention us after the new push so we can get notified and come back to review.

bnoordhuis commented on src/string_bytes.cc:
```diff
@@ -23,7 +24,6 @@ using v8::String;
 using v8::Value;
 using v8::MaybeLocal;
 
-
```
No whitespace changes please.

bnoordhuis commented on src/string_bytes.cc:
```diff
@@ -857,7 +855,16 @@ Local<Value> StringBytes::Encode(Isolate* isolate,
                                  size_t buflen) {
   Local<String> val;
-
+  uint16_t* dst = NULL;
```
Seeing how you're writing new code here, you might as well replace this with `std::vector<uint16_t>` so there is no need to explicitly free the memory.

bnoordhuis commented on src/string_bitflip.cc:
```diff
@@ -0,0 +1,20 @@
+#include "node.h"
+#include "v8.h"
+
+#include <string.h>
```
You don't need string.h here, you do need stdint.h (officially - you'll get it implicitly from node.h and v8.h but explicit > implicit.)

bnoordhuis commented on src/string_bitflip.cc:
```diff
@@ -0,0 +1,20 @@
+
+namespace node {
+
+void StringBitflip::BE_FlipBits (uint16_t* dst,
```
Style: no space before the paren.  The function name is a little off style-wise.  I'd name it `SwapBytes()` without the underscore.

bnoordhuis commented on src/string_bitflip.cc:
```diff
@@ -0,0 +1,20 @@
+    dst[i] = (src[i] << 8) | (src[i] >> 8);
+  }
+
+};
```
Style: no ; and no blank lines in the function body.  You should put a blank line _after_ the function body however. :-)

bnoordhuis replied:
Oh, another thing: I'd put this in util.h and util-inl.h.  Having it in a separate compilation unit means the compiler won't know if src and dst alias, it will always have to assume the pessimistic case.  (Maybe not true with LTO but we don't build with LTO currently.)

bnoordhuis commented:
Commit log style nit: please see CONTRIBUTING.md for how it should be written.

exinfinitum commented:
PR Updated

john-yan commented on src/string_bytes.cc:
```diff
@@ -3,9 +3,11 @@
 #include "node.h"
 #include "node_buffer.h"
 #include "v8.h"
+#include "util.h"
```
Please check to see if we need to include util.h here?

john-yan commented on src/util-inl.h:
```diff
@@ -2,6 +2,7 @@
 #define SRC_UTIL_INL_H_
 
 #include "util.h"
+#include <iostream>
```
it's not necessary to include iostream here.

john-yan commented on src/util-inl.h:
```diff
@@ -198,6 +199,14 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
   return static_cast<TypeName*>(pointer);
 }
 
+void SwapBytes(uint16_t* dst,
```
Could you make a more generic template function to work for types with 4 bytes and 8 bytes?

exinfinitum commented:
Updated

bnoordhuis commented on src/string_bytes.cc:
```diff
@@ -857,7 +856,15 @@ Local<Value> StringBytes::Encode(Isolate* isolate,
                                  size_t buflen) {
   Local<String> val;
-
+  std::vector<uint16_t> dst(buflen);
```
I'd declare an empty std::vector here and .resize() it inside the if block.  That avoids the memory allocation on little-endian architectures.

bnoordhuis commented on src/util-inl.h:
```diff
@@ -198,6 +198,25 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+  }
+}
+
 }  // namespace node
```
Can you please make sure this file passes `make lint`?

bnoordhuis commented on src/util-inl.h:
```diff
@@ -198,6 +198,25 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+               size_t buflen) {
+ 
+  //No types with size > 8 or odd sizes permitted. (This is BYTE swapping!)
+  int t_size = sizeof(TypeName);
```
Please use `const size_t` here.

bnoordhuis commented on src/util-inl.h:
```diff
@@ -198,6 +198,25 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+ 
+  //No types with size > 8 or odd sizes permitted. (This is BYTE swapping!)
+  int t_size = sizeof(TypeName);
+  static_assert(sizeof(TypeName) <= 8 && (sizeof(TypeName) & 0x1) == 0, "Unsupported type");
```
This alignment check is really only sufficient for two-byte types.

bnoordhuis commented on src/util-inl.h:
```diff
@@ -198,6 +198,25 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+  uint8_t* dst_unsigned = reinterpret_cast<uint8_t*>(dst);
+  for (size_t i = 0; i < buflen; i++) {
+    for (size_t j = 0; j < t_size; j++) {
+      dst_unsigned[t_size * i + j] = src_unsigned[t_size * i + t_size - j - 1];
```
Does this actually work when `src == dst`?

exinfinitum commented:
Updated

john-yan commented on src/util-inl.h:
```diff
@@ -198,6 +198,31 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+  static_assert(sizeof(TypeName) <= 8
+  && (sizeof(TypeName) & 0x1) == 0, "Unsupported type");
+
+  // reinterpret src as unsigned because we want unsigned shifts
```
there is no shift operation anymore :)

trevnorris commented on src/util-inl.h:
```diff
@@ -198,6 +198,31 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+  // No types with size > 8 or odd sizes (not 2-byte aligned) permitted.
+  const size_t t_size = sizeof(TypeName);
+  static_assert(sizeof(TypeName) <= 8
+  && (sizeof(TypeName) & 0x1) == 0, "Unsupported type");
```
does the linter complain here?

exinfinitum replied:
No linter errors here (or elsewhere in the code), at least from my `make lint` tests.

trevnorris commented on src/string_bytes.cc:
```diff
@@ -857,7 +856,16 @@ Local<Value> StringBytes::Encode(Isolate* isolate,
+    // http://nodejs.org/api/buffer.html regarding Node's "ucs2"
+    // encoding specification
+    dst.resize(buflen);
+    SwapBytes(&dst[0], buf, buflen);
```
Just curious, since we know `buflen` before allocating, and I don't see any vector specific operations besides `.resize()`, is using `std::vector` necessary?

john-yan replied:
I don't think it's necessary. Simple "new uint16_t[buflen]" is sufficient, and using a vector would just increase overhead. Do you think so? @bnoordhuis

exinfinitum replied:
Use of `std::vector` was suggested to me by bnoordhuis to avoid explicitly freeing memory. See https://github.com/nodejs/node/pull/3410#discussion-diff-42548236.

trevnorris commented:
@bnoordhuis think a test would be useful that writes a string and checks byte order based on platform?

trevnorris commented on src/util-inl.h:
```diff
@@ -198,6 +198,31 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+  const uint8_t* src_unsigned = reinterpret_cast<const uint8_t*>(src);
+  uint8_t* dst_unsigned = reinterpret_cast<uint8_t*>(dst);
+  for (size_t i = 0; i < buflen; i++) {
+    for (size_t j = 0; j < t_size/2; j++) {
```
nit: `t_size / 2;`

thought the linter would catch this. bad assumption.

trevnorris commented:
code's looking great. thanks!

exinfinitum commented:
Updated

trevnorris commented on src/util.h:
```diff
@@ -176,6 +176,11 @@ inline void ClearWrap(v8::Local<v8::Object> object);
+template <typename TypeName>
+inline void SwapBytes(TypeName* dst,
+                      const TypeName* src,
+                      size_t buflen);
```
style nit: this looks like it could fit under 80 chars on one line.

trevnorris commented on src/util-inl.h:
```diff
@@ -198,6 +198,35 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+template <typename TypeName>
+void SwapBytes(TypeName* dst,
+               const TypeName* src,
+               size_t buflen) {
```
ditto on the one line.

trevnorris commented:
two style nits, but LGTM otherwise.

exinfinitum commented:
@bnoordhuis @trevnorris @jasnell Wondering if I should keep the vector dst (see https://github.com/nodejs/node/pull/3410/files#r42901051), or change it back to a heap-allocated array?

bnoordhuis commented on src/util-inl.h:
```diff
@@ -198,6 +198,35 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+    break;
+    }
+  }
+}
```
This is much too obtuse to my liking.  I'd just restrict it to a non-template version that does only `uint16_t` and keeps the simple byte swapping logic.

bnoordhuis commented:
> Wondering if I should keep the vector dst

I say yes.  Using STL containers is a pattern I want to introduce more over time to make it harder to introduce memory leaks.

I think the overhead argument that @john-yan voiced is mostly a red herring at -O2.  The compiler has perfect visibility, it should have no trouble optimizing away the memory zeroing in the resize() call.

exinfinitum commented:
@bnoordhuis @jasnell @trevnorris Updated: Changed to a uint16_t only non-template function, and added support for built-in byte swapping function in the GCC compiler.

bnoordhuis commented on src/util-inl.h:
```diff
@@ -198,6 +198,18 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+void SwapBytes(uint16_t* dst, const uint16_t* src, size_t buflen) {
+  for (size_t i = 0; i < buflen; i++) {
+      #ifdef __GNUC__
+        dst[i] = __builtin_bswap16(src[i]);
```
There's really no need for this.  Check the generated machine code with and without the intrinsic.

john-yan replied:
Hello @bnoordhuis , for x86, this indeed generates exact same instruction, but x86 won't run this code because it's a little endian arch. However, for big endian ones, this could make a difference.

john-yan replied:
@bnoordhuis  system z might have dedicate byte swap machine instructions.

bnoordhuis replied:
What I mean is that compilers are generally very good at recognizing these kinds of operations and lowering them to whatever dedicated instructions are available.

john-yan replied:
it really depends on what optimization level is chosen and what compiler is using. gcc 4.8 with -O1 is not be able to lowering the shifting operation to a swap. Here is the example:
```
#include <stdio.h>

void test1(unsigned short *a) {
    *a = (*a >> 8) | (*a << 8);
}

void test2(unsigned short *b) {
    *b = __builtin_bswap16(*b);
}

int main() {
    short a = 4;
    short b = 5;
    test1(&a);
    test2(&b);
    printf("%d %d", (int)a, (int)b);
}
```
compile with "gcc -O1 --no-inline test.c" on a PPC machine
```
(gdb) disas test1
Dump of assembler code for function .test1:
   0x0000000010000604 <+0>:     lhz     r9,0(r3)
   0x0000000010000608 <+4>:     rlwinm  r10,r9,8,0,23
   0x000000001000060c <+8>:     rlwinm  r9,r9,24,8,31
   0x0000000010000610 <+12>:    or      r9,r10,r9
   0x0000000010000614 <+16>:    sth     r9,0(r3)
   0x0000000010000618 <+20>:    blr
   0x000000001000061c <+24>:    .long 0x0
   0x0000000010000620 <+28>:    .long 0x0
   0x0000000010000624 <+32>:    .long 0x0
End of assembler dump.
(gdb) disas test2
Dump of assembler code for function .test2:
   0x0000000010000628 <+0>:     lhz     r9,0(r3)
   0x000000001000062c <+4>:     sthbrx  r9,0,r3
   0x0000000010000630 <+8>:     blr
   0x0000000010000634 <+12>:    .long 0x0
   0x0000000010000638 <+16>:    .long 0x0
   0x000000001000063c <+20>:    .long 0x0
End of assembler dump.

```

bnoordhuis replied:
What about -O2?  FWIW, gcc 5.1.1 on x86_64 emits a single xchg.

john-yan replied:
in this simple case, test1 with -O2 produces the same instruction as test2 with -O1. And test2 is the same.

bnoordhuis replied:
Okay, can you add a comment explaining that __builtin_bswap16 generates more efficient code with g++ 4.8 on ppc etc.?

Also, can you not indent the #ifdef directives?

bnoordhuis commented:
@exinfinitum For extra karma maybe add a few regression tests to `test/cctest/util.cc`.

exinfinitum commented:
WARNING: Regression detected in tests: tests/parallel/test-stringbytes-external, tests/parallel/test-buffer, tests/parallel/test-buffer-indexof, and tests/parallel/test-string-decoder. The cause of this regression has been identified and will be corrected in the next update.

The cause is the presence of another byte swapping operation in the "StringSlice<UCS2>" function in src/node_buffer.cc, which performs a byte swap before executing StringBytes::Encode. The result is that two byte swaps are performed.

exinfinitum commented:
@bnoordhuis @jasnell @trevnorris Updated. Fixed regression.

john-yan commented on src/util-inl.h:
```diff
@@ -198,6 +198,20 @@ TypeName* Unwrap(v8::Local<v8::Object> object) {
+  for (size_t i = 0; i < buflen; i++) {
+      // __builtin_bswap16 generates more efficient code with
+      // g++ 4.8 on PowerPC and other big-endian archs
+      #ifdef __GNUC__
```
don't indent "#ifdef" please

bnoordhuis commented:
@exinfinitum Can you fix the style issue @john-yan pointed out and squash the commits?

CI: https://ci.nodejs.org/job/node-test-pull-request/651/

exinfinitum commented:
@bnoordhuis @john-yan Squashed.

bnoordhuis commented on src/node_buffer.cc:
```diff
@@ -460,10 +460,11 @@ void StringSlice<UCS2>(const FunctionCallbackInfo<Value>& args) {
-  if (IsLittleEndian() && aligned) {
-    buf = reinterpret_cast<const uint16_t*>(data);
-  } else {
+  if (IsLittleEndian() && !aligned) {
```
Why this change?  It looks like a semantic null change and it makes the diff noisier than it needs to be.

exinfinitum replied:
@bnoordhuis 
In this file, there are four cases that must be considered:
1. Little endian machine, properly aligned buf
2. Little endian machine, misaligned buf
3. Big endian machine, properly aligned buf
4. Big endian machine, misaligned buf

At present, reinterpreting "buf" as data is only done for the first case, and all other cases (2, 3 and 4) will involve a copy being made to produce an aligned version of unaligned data. This operation has the side effect of doing endian swapping ONLY when run on a big-endian system. This means that on a big-endian system, we will have data stored for the wrong endianness (little endian data gets swapped by alignment operation to big endian, then encode runs its own swapping function which swaps it to little endian again).

The proposed change will apply the alignment correction operation only for case 2 (little endian, misaligned). Case 1 (little endian, aligned) needs neither endianness swapping or alignment correction, while cases 3 and 4 will be handled using StringBytes::Encode's swap operation (which also takes care of alignment issues).

ALL cases need that buf is reinterpreted as a uint16_t variable, so we reinterpret either way.

bnoordhuis replied:
Okay, that makes sense.  Thanks.

bnoordhuis commented:
LGTM.  CI: https://ci.nodejs.org/job/node-test-pull-request/680/

exinfinitum commented:
@bnoordhuis Have looked at the CI test results, seems like one failure due to flaky test and one failure that does not seem to be specifically related to this change. What are the next steps to get this change integrated?

bnoordhuis commented:
@exinfinitum Sorry about that.  Landed in 2ccde01, thanks.  I changed to the Author field to your real name, you were showing up as "exinfinitum".

cjihrig commented:
After rebasing, this commit is giving me a compile failure:

```
../src/util-inl.h:206:14: error: use of undeclared identifier
      '__builtin_bswap16'
    dst[i] = __builtin_bswap16(src[i]);
```

exinfinitum commented:
@cjihrig What compiler/platform are you using?

cjihrig commented:
Thanks @exinfinitum. I had a lot of updates to install, but it works now. Strange that this is the first issue I had run into with having an older compiler.

mhdawson commented:
I think this needs to go into the LTS stream as it causes a functional issue on big endian platforms

mhdawson commented:
Reopened until this is pulled into LTS and earlier releases

TheAlphaNerd commented:
@mhdawson fwiw we have been keeping issues closed and using the lts-watch tags to figure out which commits need to land into LTS. We have actually been waiting specifically for issues to be closed before doing so to know they have landed on master.

mhdawson commented:
ah ok, understood closing again

rvagg commented:
this seems semver-minor to me, anyone disagree?

jasnell commented:
+1
On Dec 4, 2015 8:57 PM, "Rod Vagg" <notifications@github.com> wrote:

> this seems semver-minor to me, anyone disagree?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/node/pull/3410#issuecomment-162139946>.
>

bnoordhuis commented:
I'd say semver-patch.  It's a bug fix and it doesn't touch any public headers.

TheAlphaNerd commented:
@jasnell can this land into 4.x as a patch or is it semver minor? Do we want to stage minor commits or just leave them tagged and port them when we are ready to cut a minor?

bnoordhuis commented on src/string_bytes.cc:
```diff
@@ -406,9 +407,7 @@ size_t StringBytes::Write(Isolate* isolate,
-        for (size_t i = 0; i < nchars; i++)
-          dst[i] = dst[i] << 8 | dst[i] >> 8;
-        break;
+        SwapBytes(dst, dst, nchars);
```
It was pointed out in #7645 that this change introduces a bug: the `break` statement should have been retained because now the bytes are swapped back by the fallback path below.

john-yan replied:
Thanks for pointing out! Does anyone already develop a fix? if not, I can do the favour.

bnoordhuis replied:
#7645 should fix this when it lands but I've made a note in case that PR stalls.
